src/SourceBuild/content/build.sh

```diff
@@ -28,68 +38,119 @@ packagesRestoredDir="${packagesDir}restored/"
-            echo "Unrecognized argument '$1'"
-            usage
-            exit 1
-            ;;
-    esac
-    shift
+  lowerI="$(echo "$1" | awk '{print tolower($0)}')"
+  case $lowerI in
+    --clean-while-building)
+      MSBUILD_ARGUMENTS+=( "-p:CleanWhileBuilding=true")
+      ;;
+    --online)
+      MSBUILD_ARGUMENTS+=( "-p:BuildWithOnlineSources=true")
+      ;;
+    --poison)
+      MSBUILD_ARGUMENTS+=( "-p:EnablePoison=true")
+      ;;
+    --run-smoke-test)
+      alternateTarget=true
+      runningSmokeTests=true
+      MSBUILD_ARGUMENTS+=( "-t:RunSmokeTest" )
+      ;;
+    --source-repository)
+      sourceRepository="$2"
+      shift
+      ;;
+    --source-version)
+      sourceVersion="$2"
+      shift
+      ;;
+    --release-manifest)
+      releaseManifest="$2"
+      shift
+      ;;
+    --use-mono-runtime)
+      MSBUILD_ARGUMENTS+=( "/p:SourceBuildUseMonoRuntime=true" )
+      ;;
+    --with-packages)
+      CUSTOM_PACKAGES_DIR="$(cd -P "$2" && pwd)"
+      if [ ! -d "$CUSTOM_PACKAGES_DIR" ]; then
+          echo "Custom prviously built packages directory '$CUSTOM_PACKAGES_DIR' does not exist"
+          exit 1
+      fi
+      shift
+      ;;
+    --with-sdk)
+      CUSTOM_SDK_DIR="$(cd -P "$2" && pwd)"
+      if [ ! -d "$CUSTOM_SDK_DIR" ]; then
+          echo "Custom SDK directory '$CUSTOM_SDK_DIR' does not exist"
+          exit 1
+      fi
+      if [ ! -x "$CUSTOM_SDK_DIR/dotnet" ]; then
+          echo "Custom SDK '$CUSTOM_SDK_DIR/dotnet' does not exist or is not executable"
+          exit 1
+      fi
+      shift
+      ;;
+    --)
+      shift
+      echo "Detected '--': passing remaining parameters '$@' as build.sh arguments."
+      break
+      ;;
+    '-?'|-h|--help)
+      print_help
+      exit 0
+      ;;
+    *)
+      echo "Unrecognized argument '$1'"
+      print_help
+      exit 1
+      ;;
+  esac
+  shift
 done
 
+GIT_DIR="$SCRIPT_ROOT/.git"
+if [ -f "$GIT_DIR/index" ]; then # We check for index because if outside of git, we create config and HEAD manually
+  if [ -n "$sourceRepository" ] || [ -n "$sourceVersion" ] || [ -n "$releaseManifest" ]; then
+    echo "ERROR: $SCRIPT_ROOT Source Link arguments cannot be used in a git repository"
+    exit 1
+  fi
+else
+  if [ -z "$releaseManifest" ]; then
+    if [ -z "$sourceRepository" ] || [ -z "$sourceVersion" ]; then
+      echo "ERROR: $SCRIPT_ROOT is not a git repository, either --release-manifest or --source-repository and --source-version must be specified"
+      exit 1
+    fi
+  else
+    if [ -n "$sourceRepository" ] || [ -n "$sourceVersion" ]; then
+      echo "ERROR: $SCRIPT_ROOT --release-manifest cannot be specified together with --source-repository and --source-version"
+      exit 1
+    fi
+
+    get_property() {
+      local json_file_path="$1"
+      local property_name="$2"
+      grep -oP '(?<="'$property_name'": ")[^"]*' "$json_file_path"
+    }
+
+    sourceRepository=$(get_property "$releaseManifest" sourceRepository)
```

Comment:
What is the UX here is a property is not specified or the value is empty?

Reply:
I made it very explicit what is missing now